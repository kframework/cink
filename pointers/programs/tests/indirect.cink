/*
  This example tests the semantics of the indirect operator.
*/

int main() {
  int x, *y, **z, u, *v;
  y = &x;
  *y = 10;
  cout << x << "\n";
  cout << *y << "\n"; 
  z = &y;
  **z = 20;
  u = **z;
  cout << u << "\n";
  cout << **z << "\n";
  v = *z;
  cout << *v << "\n"; 
}
