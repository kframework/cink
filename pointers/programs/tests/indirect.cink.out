10
10
20
20
20
<T>
    <in>
        ListItem(#buffer ( "" ))
        ListItem(#istream ( 0 ))
    </in>
    <out>
        ListItem(#ostream ( 1 ))
        ListItem(#buffer ( "" ))
    </out>
    <control>
        <k>
            .K
        </k>
        <return>
            noVal
        </return>
        <fstack>
            .List
        </fstack>
        <sideEffects>
            .Bag
        </sideEffects>
        <env>
            main |-> #symScalLoc(1)
        </env>
    </control>
    <genv>
        main |-> #symScalLoc(1)
    </genv>
    <store>
        #symScalLoc(1) |-> lambda ( .Decls @ ((((((((((((((int x) ;) ((y 
          of (pointer to int) ;) ((z of (pointer to (pointer to int)) ;) (((
          int u) ;) (v of (pointer to int) ;))))) ((y = (& x)) ;)) (((* y) = 
          10) ;)) (((cout << x) << "\n") ;)) (((cout << (* y)) << "\n") ;)) ((
          z = (& y)) ;)) (((* (* z)) = 20) ;)) ((u = (* (* z))) ;)) (((cout <<
           u) << "\n") ;)) (((cout << (* (* z))) << "\n") ;)) ((v = (* z)) ;))
           (((cout << (* v)) << "\n") ;)) )
        #symScalLoc(2) |-> 20
        #symScalLoc(3) |-> #symScalLoc(2)
        #symScalLoc(4) |-> #symScalLoc(3)
        #symScalLoc(5) |-> 20
        #symScalLoc(6) |-> #symScalLoc(2)
    </store>
</T>
